### PR TITLE
make: introduce netdev_default pseudomodule

### PR DESCRIPTION
**Pull request by OlegHahm**
Additionally the dependencies for GNRC specific modules are centralized in Makefile.dep.

Depends on #4646.

### CODE REVIEW

**Comment by authmillenon** [comment]
Why the dependency to #4646?

**Comment by OlegHahm** [comment]
Because otherwise I would have to update it afterwards. ;-)

**Comment by OlegHahm** [comment]
btw. I'm not so sure if `netif_default` is a good name or somehow confusing.

**Comment by authmillenon** [comment]
Maybe `netdev2_default`?

**Comment by kaspar030** [comment]
@OlegHahm Could you summarize the module's intended use?

**Comment by OlegHahm** [comment]
The idea is similar to `gnrc_netif_default`, just for the more generic (no pun intended) case where one wants to use the default network interface of a board without GNRC.

**Comment by OlegHahm** [comment]
> Maybe netdev2_default?

Well, it's not necessarily bound to netdev2, either. (Actually, the Phytec transceiver is not yet netdev2.) In theory we could name it `netdev_default`, since `netdev2` is to be renamed soon anyway.

File: Makefile.dep [hunk 1]
```diff
@@ -12,6 +12,31 @@ ifneq (,$(filter nhdp,$(USEMODULE)))
   USEMODULE += oonf_rfc5444
 endif
 
+ifneq (,$(filter gnrc_netif_default,$(USEMODULE)))
+  USEMODULE += gnrc_netif
+  USEMODULE += netif_default
```
Comment (OlegHahm) [inline Makefile.dep]: This is probably superfluous since the `Makefile.dep` of the board is evaluated before the general `Makefile.dep` here.

Comment (authmillenon) [inline Makefile.dep]: I would vote for the `Makefile.dep` to pull in the device on `netif_default`, not `gnrc_netif_default`.

Comment (OlegHahm) [inline Makefile.dep]: Can you rephrase?

Comment (authmillenon) [inline Makefile.dep]: E.g.

```Makefile
ifneq (,$(filter netif_default,$(USEMODULE)))
    USEMODULE += at86rf231
  endif
```

instead of

E.g.

```Makefile
ifneq (,$(filter gnrc_netif_default,$(USEMODULE)))
    USEMODULE += at86rf231
  endif
```

Comment (authmillenon) [inline Makefile.dep]: (or instead of)

```Makefile
ifneq (,$(filter netif_default gnrc_netif_default,$(USEMODULE)))
    USEMODULE += at86rf231
endif
```

as your fixes do, for that matter.

Comment (OlegHahm) [inline Makefile.dep]: Ah, you mean we should drop `gnrc_netif_default` completely? Might make sense.

Comment (authmillenon) [inline Makefile.dep]: Maybe not completely, as it might make sense as a pseudo module to pull in netif_default + gnrc_networking, but at least from the board perspective it doesn't make sense to consider.

**Comment by authmillenon** [comment]
>>×Maybe netdev2_default?
>
> Well, it's not necessarily bound to netdev2, either. (Actually, the Phytec transceiver is not yet netdev2.) In theory we could name it netdev_default, since netdev2 is to be renamed soon anyway.

**Comment by authmillenon** [comment]
(the mobile interface of GitHub is shitty btw. One wrong tab with the thumb and the PR is closed -.-)

File: Makefile.dep (hunk 1 above)
Comment (OlegHahm) [inline Makefile.dep]: Hm, if someone wants to use any GNRC module, he/she would specify the particular GNRC module anyway, right?

Comment (authmillenon) [inline Makefile.dep]: > Maybe not completely, as it might make sense as a pseudo module to pull in netif_default + gnrc_networking, but at least from the board perspective it doesn't make sense to consider.

I meant `gnrc_netif`, not `gnrc_networking`.

**Comment by authmillenon** [comment]
>> Maybe netdev2_default?
>
> Well, it's not necessarily bound to netdev2, either. (Actually, the Phytec transceiver is not yet netdev2.) In theory we could name it netdev_default, since netdev2 is to be renamed soon anyway.

This should have been edited above, but somehow the mobile interface did not let it through either... So for clarity again:

To me, `netdev` is a unifying term for both `gnrc_netdev` and `netdev2`, while `netif` is something that should sit on top of a `netdev` and may have some MAC implementation in between. So calling the pseudo-module `netif_default` is definitely wrong in my opinion, because it only pulls in the `netdev` => I would vote for calling it `netdev_default` right away.

**Comment by kaspar030** [comment]
> I would vote for calling it netdev_default right away.

For different reasons, but +1!

File: Makefile.dep (hunk 1 above)
Comment (OlegHahm) [inline Makefile.dep]: Can you describe a usecase where someone wants to add `gnrc_netif_default` without any other GNRC module?

Comment (authmillenon) [inline Makefile.dep]: No, that is why it doesn't make any sense at the board level. That's what I'm saying. But it DOES make sense to give the user the option to say

```Makefile
USEMODULE += gnrc_netdev_default
```

instead of

```Makefile
USEMODULE += netdev_default
USEMODULE += gnrc_netdev2 # or whatever gnrc_netdev requires
```

at the application level.

Comment (OlegHahm) [inline Makefile.dep]: Yes, but in which case does an application require `gnrc_netdev2` without any other GNRC module?

Comment (authmillenon) [inline Makefile.dep]: `examples/default` e.g.

Comment (OlegHahm) [inline Makefile.dep]: Hm, okay, I see, we need it for now, but theoretically things like the default example should not require any GNRC module.

Comment (authmillenon) [inline Makefile.dep]: Yes :-)

Comment (OlegHahm) [inline Makefile.dep]: Hm, okay, now back to your original proposal. If I do:
```
ifneq (,$(filter netif_default,$(USEMODULE)))
    USEMODULE += at86rf231
endif
```
in the board makefile like I originally had, who should pull in the `gnrc_netif_default` then?

Comment (authmillenon) [inline Makefile.dep]: noone. `Makefile.dep` in `RIOT_BASE` should pull in `netif_default` + whatever else is required to get GNRC running if `gnrc_netif_default` is included.

Comment (OlegHahm) [inline Makefile.dep]: But isn't the line we're discussing here doing exactly that?

Comment (authmillenon) [inline Makefile.dep]: Yes! This is why it should not be removed, but the dependency of `gnrc_netif_default` in `boards/*/Makefile.dep` should be! :-)

Comment (OlegHahm) [inline Makefile.dep]: But if include `gnrc_netif_default` in my application's Makefile, the board wouldn't add the correct driver to the modules then.

Comment (authmillenon) [inline Makefile.dep]: Why?

Comment (OlegHahm) [inline Makefile.dep]: How would it know that it should?

Comment (authmillenon) [inline Makefile.dep]: Clarified offline. Now I see that it only is possible the way it is currently implemented.

**Comment by OlegHahm** [comment]
Rebased to master after #4646 got merged and renamed to `*netdev_default`.

**Comment by authmillenon** [comment]
Tests seem to be working still, let's see what Murdock is saying. Please squash.

**Comment by OlegHahm** [comment]
Squashed

**Comment by authmillenon** [comment]
You removed the `NEEDS SQUASHING` label to late for Murdock ;-)

**Comment by authmillenon** [comment]
And go.
